torch/torch_version.py

```diff
@@ -0,0 +1,68 @@
+from functools import total_ordering
+from typing import Iterable, Union
+
+from pkg_resources import packaging  # type: ignore[attr-defined]
+
+Version = packaging.version.Version
+InvalidVersion = packaging.version.InvalidVersion
+
+from .version import __version__ as internal_version
+
+
+@total_ordering
+class TorchVersion(str):
+    """A string with magic powers to compare to both Version and iterables!
+
+    Prior to 1.10.0 torch.__version__ was stored as a str and so many did
+    comparisons against torch.__version__ as if it were a str. In order to not
+    break them we have TorchVersion which masquerades as a str while also
+    having the ability to compare against both packaging.version.Version as
+    well as tuples of values, eg. (1, 2, 1)
+
+    Examples:
+        Comparing a TorchVersion object to a Version object
+            TorchVersion('1.10.0a') > Version('1.10.0a')
+
+        Comparing a TorchVersion object to a Tuple object
+            TorchVersion('1.10.0a') > (1, 2)    # 1.2
+            TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
+
+        Comparing a TorchVersion object against a string
+            TorchVersion('1.10.0a') > '1.2'
+            TorchVersion('1.10.0a') > '1.2.1'
+    """
+    # fully qualified type names here to appease mypy
+    def _convert_to_version(self, inp: Union[packaging.version.Version, str, Iterable]) -> packaging.version.Version:
```

Comment:
As you are creating shortcut on line 7, referencing version by full name is no longer neeeded, is it?
```suggestion
    def _convert_to_version(self, inp: Union[Version, str, Iterable]) -> Version:
```

Reply:
Unfortunately I had to use the fully qualified name here since `mypy` didn't like it when I used the short hand version

Reply:
Without this mypy actually fails see:
```
.venv ❯ mypy --config=mypy.ini torch/torch_version.py
torch/torch_version.py:35:46: error: Variable "torch.torch_version.Version" is not valid as a type  [valid-type]
torch/torch_version.py:35:46: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
Found 1 error in 1 file (checked 1 source file)
```

<details>

<summary> Diff </summary>

```diff
diff --git a/torch/torch_version.py b/torch/torch_version.py
index bb01294aeb..e987452711 100644
--- a/torch/torch_version.py
+++ b/torch/torch_version.py
@@ -32,7 +32,7 @@ class TorchVersion(str):
             TorchVersion('1.10.0a') > '1.2.1'
     """
     # fully qualified type names here to appease mypy
-    def _convert_to_version(self, inp: Union[packaging.version.Version, str, Iterable]) -> packaging.version.Version:
+    def _convert_to_version(self, inp: Union[Version, str, Iterable]) -> Version:
         if isinstance(inp, Version):
             return inp
         elif isinstance(inp, str):
```

</details>

Context: https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases